### PR TITLE
Signature Insights

### PR DESCRIPTION
**Pull request by Montoya**
Draft proposal for a new signature insights API based on transaction insights

### CODE REVIEW

File: SIPS/sip-16.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+```typescript
+import { OnSignatureHandler } from "@metamask/snap-types";
+
+enum SeverityLevel {
+  Critical = 'critical',
+}
+
+export const onSignature: OnSignatureHandler = async ({
+  signature,
```
Comment (FrederikBolding) [inline SIPS/sip-16.md:99]: The parameters here should be typed out.

File: SIPS/sip-16.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+```
+
+### Snap Implementation
+
+The following is an example implementation of the API:
+
+```typescript
+import { OnSignatureHandler } from "@metamask/snap-types";
+
+enum SeverityLevel {
```
Comment (FrederikBolding) [inline SIPS/sip-16.md:94]: Import from `snaps-types` instead seems clearer

File: SIPS/sip-16.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
+---
+sip: 16
+title: Signature Insights
+status: Draft
+discussions-to: https://github.com/MetaMask/SIPs/discussions/32
```
Comment (FrederikBolding) [inline SIPS/sip-16.md:5]: This link seems wrong

File: SIPS/sip-16.md (hunk 1 above)
Comment (Montoya) [reply SIPS/sip-16.md:99]: Oh good point. Does that mean this docs example needs to be typed out too? https://docs.metamask.io/snaps/reference/exports/#example-1

File: SIPS/sip-16.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+### Definitions
+
+> This section is non-normative, and merely recapitulates some definitions from [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md).
+
+- `ChainId` - a [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md) string.
+  It identifies a specific chain among all blockchains recognized by the CAIP standards. The ChainId refers to the network that the signing account is connected to. It is not necessarily the network where the signed message will be broadcast. 
+  - `ChainId` consists of a `Namespace` and a `Reference`
+    - `Namespace` - a class of similar blockchains. For example EVM-based blockchains.
+    - `Reference` - a way to identify a concrete chain inside a `Namespace`. For example Ethereum Mainnet or one of its test networks.
+- `Method` - a string which can be any of the following 5 [signing methods](https://docs.metamask.io/wallet/concepts/signing-methods/) supported by the MetaMask wallet API. The signing method is required by MetaMask for the dapp to make a signature request. It is not part of the signature payload.
```
Comment (hmalik88) [inline SIPS/sip-16.md:51]: It might be worth it to mention that this list would be expanded with custom signature types in the future.

Comment (Montoya) [reply SIPS/sip-16.md]: Added a sentence in latest commit

File: SIPS/sip-16.md (hunk 1 above)
Comment (Montoya) [reply SIPS/sip-16.md]: I added types. The only one that isn't obvious is the signature payload. I used `any`.

File: SIPS/sip-16.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+```typescript
+import { OnSignatureHandler } from "@metamask/snaps-types";
+
+enum SeverityLevel {
+  Critical = 'critical',
+}
+
+export const onSignature: OnSignatureHandler = async ({
+  signature: any,
+  method: string,
```
Comment (hmalik88) [inline SIPS/sip-16.md:100]:
```suggestion
```
This can live within the signature object.

File: SIPS/sip-16.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  chainId: string,
+  signatureOrigin: string, /* If allowSignatureOrigin is set to true */
```
Comment (hmalik88) [inline SIPS/sip-16.md:102]:
```suggestion
  origin: string, /* If allowSignatureOrigin is set to true */
```
I think we're overstating, I think it's understood what this origin is for.

File: SIPS/sip-16.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+The permission includes an OPTIONAL caveat `allowSignatureOrigin`. 
+The caveat grants a Snap read-only access to the URL requesting the signature.
+It can be specified as follows: 
+
+```json
+{
+  "initialPermissions": {
+    "endowment:signature-insight": {
+      "allowSignatureOrigin": true
```
Comment (hmalik88) [inline SIPS/sip-16.md:81]:
```suggestion
      "allowOrigin": true
```

see below.

File: SIPS/sip-16.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+```json
+{
+  "initialPermissions": {
+    "endowment:signature-insight": {}
+  }
+}
+```
+
+The permission includes an OPTIONAL caveat `allowSignatureOrigin`.
```
Comment (hmalik88) [inline SIPS/sip-16.md:73]:
```suggestion
The permission includes an OPTIONAL caveat `allowOrigin`. 
```

Comment (Montoya) [reply SIPS/sip-16.md:73]: OK but the caveat for transaction insights is `allowTransactionOrigin` so for consistency I figured this one would be `allowSignatureOrigin`

File: SIPS/sip-16.md (hunk 6 above)
Comment (Montoya) [reply SIPS/sip-16.md:102]: The transactionOrigin handler is like this: 

```
export const onTransaction: OnTransactionHandler = async ({
  transaction,
  chainId,
  transactionOrigin,
}) => {
```

File: SIPS/sip-16.md (hunk 8 above)
Comment (hmalik88) [reply SIPS/sip-16.md:73]: Ok cool

File: SIPS/sip-16.md (hunk 2 above)
Comment (hmalik88) [reply SIPS/sip-16.md:94]: Why was this resolved? Was there further conversation about this?

Comment (hmalik88) [reply SIPS/sip-16.md]: I think you might have misunderstood what Frederik meant, I believe he was referring to importing the enum. It was then the right package, but now we use the sdk so no longer valid lol. Nonetheless, I've fixed in [6da2614](https://github.com/MetaMask/SIPs/pull/117/commits/6da261452194c304f43c28a220dfa55e26cf17ee)

File: SIPS/sip-16.md [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+interface SignTypedDataV4SignatureObj {
+  from: string;
+  data: Record<string, any>;
+  metamaskId: string;
+  signatureMethod: 'eth_signTypedData_v4';
+  version: 'V4';
+}
+```
+
+**Note**: The `metamaskId`, `signatureMethod` and `version` properties are MetaMask specific properties and not reflective of the standards defining the underlying signature methods. `signatureMethod` SHOULD be used by the signature insight snap as the source of the truth to identify the signature scheme it is providing insights for.
```
Comment (FrederikBolding) [inline SIPS/sip-16.md:205]: Wondering if we should strip `metamaskId` and `version` from the payload? It seems like noise that no one will need?

File: SIPS/sip-16.md [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+Please see [SIP-11](sip-3.md) for more information on Transaction insight severity levels.
+
+## Appendix I: Ethereum Signature Objects
+
+The following signature objects may appear for any `chainId` of `eip155:*` where `*` is some positive integer. This includes all Ethereum or "EVM-compatible" chains. As of the time of the creation of this SIP, they are the only possible signature objects for Ethereum chains.
+
+### eth_sign
+
+```typescript
+interface EthSignatureObj {
```
Comment (FrederikBolding) [inline SIPS/sip-16.md:151]: Can we rename all of these?
```suggestion
interface EthSignature {
```

File: SIPS/sip-16.md [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+
+### Snap Implementation
+
+The following is an example implementation of the API:
+
+```typescript
+import { OnSignatureHandler, SeverityLevel } from "@metamask/snaps-sdk";
+
+export const onSignature: OnSignatureHandler = async ({
+  signature: any,
```
Comment (FrederikBolding) [inline SIPS/sip-16.md:95]:
```suggestion
  signature: Record<string, unknown>,
```

File: SIPS/sip-16.md (hunk 9 above)
Comment (hmalik88) [reply SIPS/sip-16.md:205]: Hmm yeah, I suppose so.

**Review by FrederikBolding — APPROVED** [review]

File: SIPS/sip-16.md [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+
+```typescript
+interface SignTypedDataV4Signature {
+  from: string;
+  data: Record<string, any>;
+  signatureMethod: 'eth_signTypedData_v4';
+}
+```
+
+**Note**: The `signatureMethod` property is MetaMask specific and not reflective of the standards defining the underlying signature methods. `signatureMethod` SHOULD be used by the signature insight snap as the source of the truth to identify the signature scheme it is providing insights for.
```
Comment (david0xd) [inline SIPS/sip-16.md:193]:
```suggestion
**Note**: The `signatureMethod` property is MetaMask specific and not reflective of the standards defining the underlying signature methods. `signatureMethod` SHOULD be used by the signature insight Snap as the source of the truth to identify the signature scheme it is providing insights for.
```

**Review by david0xd — APPROVED** [review]

**Pull request merged**